### PR TITLE
Autogenerate routes if host is not specified - based on service and namespace ala OS v{1,2}

### PR DESCRIPTION
**Pull request by ramr**
Autogenerate host names (if none are specified) for routes based on service and namespace ala 
openshift v{1,2}.

@pweil-  PTAL
Let me know if I got this right, still a newb.  There's probably some additional work to pass a config variable/value down to the API - still figuring that out - probably via kubeconfig would be the right answer, I dunno (will chat w/ you on Monday). Thanks.

Unit tests pass (make check) and integration tested this with the sample app https://github.com/ramr/origin/tree/master/examples/sample-app   - removing the "host" defined for the frontend in examples/sample-app/application-template-stibuild.json  and getting one to autogenerate based on the service + namespace. 

```    
[vagrant@openshiftdev sample-app]$ osc get -n test routes    
NAME                HOST/PORT                        PATH                SERVICE             LABELS    
route-edge          frontend-test.v3.openshift.com                       frontend            template=application-template-stibuild    
```

### CODE REVIEW

File: pkg/route/api/allocator/allocator.go [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+// Would be better if we could use "v3.openshift.app", someone bought that!
+const DEFAULT_DNS_SUFFIX = "v3.openshift.com"
+
+// Invalid host name characters or not the <let-or-digit-or-hyphen> and
+// '.' separator as defined in rfc1123 (via rfc952)
+var name_RE, _ = regexp.Compile("[^A-Za-z0-9-.]+")
+
+func allocateShard(route *routeapi.Route) string {
+	//  For later, allocate names pointing to different shards.
+	shard := DEFAULT_DNS_SUFFIX
```
Comment (smarterclayton) [inline pkg/route/api/allocator/allocator.go]: Allocation (once we have shards) will be a controller and a separate process.

Turn this into a type, take this as a value passed in, and then do it that way.

File: pkg/route/api/allocator/allocator.go [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+	}
+
+	var s = ""
+	if len(route.Namespace) > 0 {
+		s = fmt.Sprintf("%s-%s.%s", name, route.Namespace, suffix)
+	} else {
+		s = fmt.Sprintf("%s.%s", name, suffix)
+	}
+
+	return name_RE.ReplaceAllString(s, "")
```
Comment (smarterclayton) [inline pkg/route/api/allocator/allocator.go]: Names should be checked up front on validation - they shouldn't get in here in an illegal form so you shouldn't have to worry about them.

File: pkg/route/api/allocator/allocator.go [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+	"regexp"
+
+	"code.google.com/p/go-uuid/uuid"
+	routeapi "github.com/openshift/origin/pkg/route/api"
+)
+
+// This should be something we get from config but we would still need a
+// default name if nothing's passed. The first pass uses the default name.
+// Would be better if we could use "v3.openshift.app", someone bought that!
+const DEFAULT_DNS_SUFFIX = "v3.openshift.com"
```
Comment (smarterclayton) [inline pkg/route/api/allocator/allocator.go]: Constants should be in "CamelCase" if public and "camelCase" if not (since this should be removed anyway, just fyi)

File: pkg/route/api/allocator/allocator.go [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+// Invalid host name characters or not the <let-or-digit-or-hyphen> and
+// '.' separator as defined in rfc1123 (via rfc952)
+var name_RE, _ = regexp.Compile("[^A-Za-z0-9-.]+")
```
Comment (smarterclayton) [inline pkg/route/api/allocator/allocator.go]: By convention all variables are camelCase or CamelCase - no underscores.

File: pkg/route/api/allocator/allocator.go [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+	return shard
+}
+
+func generateName(route *routeapi.Route, suffix string) string {
+	name := route.ServiceName
+	if len(name) == 0 {
+		name = uuid.NewUUID().String()
+	}
+
+	var s = ""
```
Comment (smarterclayton) [inline pkg/route/api/allocator/allocator.go]: Convention is `s := ""` here.

File: pkg/route/api/allocator/allocator.go (hunk 4 above)
Comment (pweil-) [inline pkg/route/api/allocator/allocator.go]: Probably want to use `MustCompile` here.  http://golang.org/pkg/regexp/#MustCompile

File: hack/test-go.sh [hunk 6]
```diff
@@ -78,5 +78,8 @@ if [[ -n "${KUBE_COVER}" && -n "${OUTPUT_COVERAGE}" ]]; then
     fi
   done
 else
-  go test $KUBE_RACE $KUBE_TIMEOUT $KUBE_COVER "${@:2}" $test_packages
+  # Allow specific tests (e.g. TestRoute1) to be run via make.
+  # Example:
+  #   make check WHAT=pkg/route/api/validation/ GOFLAGS="-v -run TestRoute1"
+  go test ${GOFLAGS:-} $KUBE_RACE $KUBE_TIMEOUT $KUBE_COVER "${@:2}" $test_packages
```
Comment (pweil-) [inline hack/test-go.sh]: This could probably be in a separate PR, I think it stands on its own

File: pkg/route/api/allocator/allocator.go (hunk 2 above)
Comment (pweil-) [inline pkg/route/api/allocator/allocator.go]: Take a look at the ValidateRoute function.  Name generation should occur before the `kval.ValidateObjectMeta` which is what is validating names and namespaces.

Comment (ramr) [inline pkg/route/api/allocator/allocator.go]: Does this also take care of ascii vs non-ascii characters? It will cause issues otherwise (ala for UCS{2,4}).

File: pkg/route/api/allocator/allocator.go (hunk 3 above)
Comment (ramr) [inline pkg/route/api/allocator/allocator.go]: Thanks. Got to know.

File: hack/test-go.sh (hunk 6 above)
Comment (ramr) [inline hack/test-go.sh]: Ok - will do this separately.

File: pkg/route/api/allocator/allocator.go (hunk 2 above)
Comment (pweil-) [inline pkg/route/api/allocator/allocator.go]: It is basically a regex that boils down to `const DNS1123LabelFmt string = "[a-z0-9]([-a-z0-9]*[a-z0-9])?"` with some other wrappers around it.  (k8s/pkg/util/validation.go)

Comment (ramr) [inline pkg/route/api/allocator/allocator.go]: No I mean what if the service name or namespace is an i18n name?     
E.g. ```ハロworldーopenshift``

File: pkg/route/api/allocator/allocator.go (hunk 5 above)
Comment (ramr) [inline pkg/route/api/allocator/allocator.go]: Thanks. Good to know.

File: pkg/route/api/allocator/allocator.go (hunk 2 above)
Comment (pweil-) [inline pkg/route/api/allocator/allocator.go]: Discussed in IRC, recording here so it's not a hanging question.  Validation for ns and name ensures a-z0-9.  

`F0309 14:14:33.643245    1869 create.go:83] Service "hello-nginx字" is invalid: metadata.name: invalid value 'hello-nginx字': must have at most 24 characters and match regex [a-z]([-a-z0-9]*[a-z0-9])`

**Comment by ramr** [comment]
@smarterclayton @pweil-  made changes as per your comments above. PTAL

File: pkg/route/registry/route/rest.go [hunk 7]
```diff
@@ -78,6 +79,13 @@ func (rs *REST) Create(ctx kapi.Context, obj runtime.Object) (runtime.Object, er
 		route.Name = uuid.NewUUID().String()
 	}
 
+	//  shards will be eventually allocated via a separate controller.
+	shard := allocator.Allocate(route)
+
+	if len(route.Host) == 0 {
```
Comment (pweil-) [inline pkg/route/registry/route/rest.go]: I think this should occur prior to `ValidateRoute`.  That way we can ensure that whatever is generated passes the same validation as a user submitted name (`util.IsDNSSubdomain`).

File: pkg/route/api/types.go [hunk 8]
```diff
@@ -30,6 +30,12 @@ type RouteList struct {
 	Items         []Route `json:"items"`
 }
 
+// RouterShard defines a routing shard.
+type RouterShard struct {
+	ShardName string
```
Comment (pweil-) [inline pkg/route/api/types.go]: Needs comments

File: pkg/route/api/allocator/allocator.go [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+	"code.google.com/p/go-uuid/uuid"
+	routeapi "github.com/openshift/origin/pkg/route/api"
+)
+
+// This should be something we get from config but we would still need a
+// default name if nothing's passed. The first pass uses the default name.
+// Would be better if we could use "v3.openshift.app", someone bought that!
+const defaultDNSSuffix = "v3.openshift.com"
+
+func Generate(route *routeapi.Route, shard *routeapi.RouterShard) string {
```
Comment (pweil-) [inline pkg/route/api/allocator/allocator.go]: godoc comments

File: pkg/route/api/allocator/allocator.go [hunk 10]
```diff
@@ -0,0 +1,30 @@
+package allocator
```
Comment (pweil-) [inline pkg/route/api/allocator/allocator.go]: @smarterclayton is this something you think warrants an interface so it can have different implementations in the future (perhaps for different allocation strategies)?

**Comment by pweil-** [comment]
@ramr couple small things and I think it looks good unless we want to abstract an allocator object (https://github.com/openshift/origin/pull/1250#discussion_r26118813)

File: pkg/route/registry/route/rest.go (hunk 7 above)
Comment (ramr) [inline pkg/route/registry/route/rest.go]: Ok - will move it up/prior to that check.

File: pkg/route/api/allocator/allocator.go (hunk 10 above)
Comment (smarterclayton) [inline pkg/route/api/allocator/allocator.go]: Yes

Comment (ramr) [inline pkg/route/api/allocator/allocator.go]: Ok - will do. Though I suspect this may need some change when the sharding support is added.

Comment (smarterclayton) [inline pkg/route/api/allocator/allocator.go]: Essentially the interface should only expose the input (route) and the output (key to router), and then the interface would be used by the allocator (which won't be in the rest call).  In fact, allocations like this being in the rest call are deceptive and should have a clear TODO: this does not belong here

Comment (ramr) [inline pkg/route/api/allocator/allocator.go]: Ok - let me add a TODO for now. Its going to take me some time to get a handle on how to do that/how it works, so it might be safer to do it in a separate PR. 
That sound good @smarterclayton ?

**Comment by pweil-** [comment]
LGTM if we're ok with waiting on extracting the interface.  Tagging for e2e.

[test]

File: pkg/route/api/allocator/allocator.go (hunk 10 above)
Comment (smarterclayton) [inline pkg/route/api/allocator/allocator.go]: I would prefer it to be presented as an interface to consumers, even if you use the static defaultAllocator method.

Comment (ramr) [inline pkg/route/api/allocator/allocator.go]: Allrighty, let me just do the whole shebang in this one then.

Comment (ramr) [inline pkg/route/api/allocator/allocator.go]: Still muddling through ... But I am going to close this PR out and create a new one.
